js/adapt-stateful-session.js

```diff
@@ -1,227 +1,259 @@
-      // @oliverfoster: this can probably be removed in any subsequent rewrite
-      if (courseState) {
-        serializer.deserialize(courseState.slice(2).map(Number).map(String).join(''), doSynchronousPart);
-      } else {
-        doSynchronousPart();
-      }
-    },
-
-    getSessionState: function() {
-      var blockCompletion = serializer.serialize();
-      var courseComplete = Adapt.course.get('_isComplete') || false;
-      var assessmentPassed = Adapt.course.get('_isAssessmentPassed') || false;
-      Adapt.log.info(`course._isComplete: ${courseComplete}, course._isAssessmentPassed: ${assessmentPassed}, BlockCompletion: ${blockCompletion}`);
-      var courseState = SCORMSuspendData.serialize([
-        courseComplete,
-        assessmentPassed,
-        ...blockCompletion.split('').map(Number).map(Boolean)
-      ]);
-      var sessionPairs = {
-        'c': courseState,
-        'q': (this._shouldStoreResponses === true ? questions.serialize() : '')
-      };
-      return sessionPairs;
-    },
-
-    // Session In Progress
-    setupEventListeners: function() {
-      $(window).on('beforeunload unload', this._onWindowUnload);
-
-      if (this._shouldStoreResponses) {
-        this.listenTo(Adapt.components, 'change:_isSubmitted', _.debounce(this.saveSessionState.bind(this), 1));
+        const [ _isComplete, _isAssessmentPassed ] = SCORMSuspendData.deserialize(sessionPairs.c);
+        Adapt.course.set({
+          _isComplete,
+          _isAssessmentPassed
+        });
       }
-
-      if (this._shouldRecordInteractions) {
-        this.listenTo(Adapt, 'questionView:recordInteraction', this.onQuestionRecordInteraction);
+      if (sessionPairs.q && this._shouldStoreResponses) {
+        this._componentSerializer.deserialize(sessionPairs.q);
       }
+    }
 
-      this.listenTo(Adapt.blocks, 'change:_isComplete', this.onBlockComplete);
+    setupEventListeners() {
+      const debouncedSaveSession = _.debounce(this.saveSessionState.bind(this), 1);
+      this.listenTo(Adapt.data, 'change:_isComplete', debouncedSaveSession);
+      if (this._shouldStoreResponses) {
+        this.listenTo(Adapt.data, 'change:_isSubmitted change:_userAnswer', debouncedSaveSession);
+      }
       this.listenTo(Adapt, {
-        'assessment:complete': this.onAssessmentComplete,
         'app:languageChanged': this.onLanguageChanged,
+        'questionView:recordInteraction': this.onQuestionRecordInteraction,
+        'assessment:complete': this.onAssessmentComplete,
         'tracking:complete': this.onTrackingComplete
       });
-      this.listenTo(Adapt.course, 'change:_isComplete', this.saveSessionState);
-    },
+      const config = Adapt.spoor.config;
+      const advancedSettings = config._advancedSettings;
+      const shouldCommitOnVisibilityChange = (!advancedSettings ||
+          advancedSettings._commitOnVisibilityChangeHidden !== false) &&
+          document.addEventListener;
```

Comment:
Just one query – is this an IE8 hangover?

Reply:
Yes